squidpy/gr/_sepal.py

```diff
@@ -0,0 +1,341 @@
+        use_ixs=False,
+        n_jobs=n_jobs,
+        backend=backend,
+        show_progress_bar=show_progress_bar,
+    )(
+        vals=vals,
+        max_n=max_nbrs,
+        n_iter=n_iter,
+        sat=sat,
+        sat_idx=sat_idx,
+        unsat=unsat,
+        unsat_idx=unsat_idx,
+        dt=dt,
+        thres=thres,
+    )
+
+    sepal_score = pd.DataFrame(score, index=genes, columns=["sepal_score"])
+    if sepal_score["sepal_score"].isna().any():
+        logg.info("Found NaN in sepal scores, consider increase `n_iter` to a higher value.")
+    sepal_score.sort_values(by="sepal_score", ascending=False, inplace=True)
+
+    if copy:
+        logg.info("Finish", time=start)
+        return sepal_score
+
+    _save_data(adata, attr="uns", key="sepal_score", data=sepal_score, time=start)
+
+
+def _score_helper(
+    ixs: Sequence[int],
+    vals: Union[spmatrix, np.ndarray],
+    max_n: int,
+    n_iter: int,
+    sat: np.ndarray,
+    sat_idx: np.ndarray,
+    unsat: np.ndarray,
+    unsat_idx: np.ndarray,
+    dt: np.float_,
+    thres: np.float_,
+    queue: Optional[SigQueue] = None,
+) -> np.ndarray:
+
+    score = []
+    if max_n == 4:
+        fun = _laplacian_rect
+    elif max_n == 6:
+        fun = _laplacian_hex
+    else:
+        NotImplementedError("Laplacian for `max_nbrs`= {max_n} is not yet implemented.")
+    for i in ixs:
+        conc = vals[:, i].copy()
+        time_iter = _diffusion(
+            conc, fun, n_iter, sat, sat_idx, sat.shape[0], unsat, unsat_idx, conc.shape[0], dt=dt, thres=thres
+        )
+        if time_iter.shape[0] != 0:
+            score.append(0.001 * time_iter[0])
+        else:
+            score.append(np.nan)
+
+        if queue is not None:
+            queue.put(Signal.UPDATE)
+
+    if queue is not None:
+        queue.put(Signal.FINISH)
+
+    return np.array(score)
+
+
+@njit(parallel=False, fastmath=True)
+def _diffusion(
+    conc: np.ndarray,
+    laplacian: Callable[..., np.float_],
+    n_iter: int,
+    sat: np.ndarray,
+    sat_idx: np.ndarray,
+    sat_shape: int,
+    unsat: np.ndarray,
+    unsat_idx: np.ndarray,
+    conc_shape: int,
+    dt: float = 0.001,
+    D: float = 1.0,
+    thres: float = 1e-8,
+) -> np.ndarray:
+    """Simulate diffusion process on a regular graph."""
+    entropy_arr = np.zeros(n_iter)
+    ent = np.ones(n_iter + 1)
+
+    nhood = np.zeros(sat_shape)
+    weights = np.ones(sat_shape)
+
+    for i in range(n_iter):
+        for j in range(sat_shape):
+            nhood[j] = np.sum(conc[sat_idx[j]])
+        d2 = laplacian(conc[sat], nhood, weights)
+
+        dcdt = np.zeros(conc_shape)
+        dcdt[sat] = D * d2
+        conc[sat] += dcdt[sat] * dt
+        conc[unsat] += dcdt[unsat_idx] * dt
+        # set values below zero to 0
+        conc[conc < 0] = 0
+        # compute entropy
+        ent[i + 1] = _entropy(conc[sat]) / sat_shape
+        entropy_arr[i] = np.abs(ent[i + 1] - ent[i])  # estimate entropy difference
```

Comment:
There's no need to store the whole `ent` array, since it's just used to calculate the diff, which is the only thing returned.
You can save space just by saving `ent: float` and `prev_env: float`.